src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutorKeepalived.java

```diff
@@ -0,0 +1,308 @@
+                backoff);
+
+        // If no response from Execute(), use WaitExecution() in a "loop" which is implicated inside
+        // the retry block.
+        if (response == null) {
+          response = retrier.execute(
+              () -> Utils.refreshIfUnauthenticated(this::waitExecution, callCredentialsProvider),
+              backoff);
+        }
+      }
+
+      return response;
+    }
+
+    @Nullable
+    ExecuteResponse execute() throws IOException {
+      Preconditions.checkState(lastOperation == null);
+
+      try {
+        Iterator<Operation> streamOperations = executionBlockingStubSupplier.get().execute(request);
+
+        // We don't want to reset backoff for Execute() since if there is an error:
+        //   1. If happened before we received a first response, we want to ensure the retry counter
+        //      is increased and call Execute() again.
+        //   2. Otherwise, we will fallback to WaitExecution() loop.
+        return handleStreamOperations(streamOperations, /* resetBackoff */ false);
+      } catch (StatusRuntimeException e) {
+        if (lastOperation != null) {
+          // By returning null, we are going to fallback to WaitExecution() loop.
+          return null;
+        }
+        throw new IOException(e);
+      }
+    }
+
+    @Nullable
+    ExecuteResponse waitExecution() throws IOException {
+      Preconditions.checkState(lastOperation != null);
+
+      WaitExecutionRequest request = WaitExecutionRequest.newBuilder()
+          .setName(lastOperation.getName())
+          .build();
+      try {
+        Iterator<Operation> streamOperations = executionBlockingStubSupplier.get()
+            .waitExecution(request);
+        // We want to reset backoff for WaitExecution() so we can "infinitely" wait for the
+        // execution to complete as long as they are making progress (by returning response).
+        return handleStreamOperations(streamOperations, /* resetBackoff */ true);
+      } catch (StatusRuntimeException e) {
+        if (e.getStatus().getCode() == Code.NOT_FOUND) {
+          // Operation was lost on the server. Retry Execute.
+          lastOperation = null;
+          return null;
+        }
+        throw new IOException(e);
+      }
+    }
+
+    /**
+     * Process a stream of operations from Execute() or WaitExecution().
+     *
+     * @param resetBackoff reset backoff if true once received a response successfully.
+     */
+    ExecuteResponse handleStreamOperations(Iterator<Operation> streamOperations,
+        boolean resetBackoff) throws IOException {
+      try {
+        while (streamOperations.hasNext()) {
+          Operation operation = streamOperations.next();
+
+          if (resetBackoff) {
+            // Assuming the server has made progress since we received the response. Reset the
+            // backoff so that this request has a full deck of retries.
+            backoff.reset();
```

Comment:
It looks like the intent is that the backoff is reset when the server thinks that progress was made. The comment in line 210 indicates that that's the point where the client is certain that there were no errors in the last `Operation`. Why not unconditionally reset the backoff there?

Reply:
Moved the code for resting after client verified no errors in the `Operation`. However, we still need to distinct between Execute() and WaitExecution() to avoid infinite loop caused by a buggy server. For example:
1. Call Execute(), server returns the first response, and then timeout when waiting for the next. Fallback to WaitExecution().
2. Call WaitExecution(), but returned NOT_FOUND (server may already finished the execution and delete the reference). Retry with Execute().

See test case `executeRemotely_notFoundLoop_reportError`.

Reply:
FWIW, if this is necessary, I think it's nicer to do it with a Boolean flag and not a lambda (lambdas make the code flow harder to follow visually). It looks like the `lastOperation = operation; observer.onNext(operation);`

So you want to protect against the case where the server "forgets" about the existence of a request between an  `Execute()` call and the subsequent `WaitExecution()` one? That makes sense, but are we then not afraid of this scenario:

1. Client sends `Execute()`, then the server breaks the stream
2. Client sends `WaitExecution()`, server returns a few `Operation` messages, then stops responding
3. GOTO 1. The backoff is reset, so this is an infinite loop.

Reply:
> 3. GOTO 1. The backoff is reset, so this is an infinite loop.

If server stops responding, we won't GOTO 1. The cases to exit the `WaitExecution()` loop:
1. Received the final response.
2. Received a noretriable gRPC error.
3. Received NOT_FOUND error where we will retry Execute().
4. Received consecutive retriable gRPC errors (up to `--remote_retry`).

Then the scenario could be:
1. Client sends `Execute()`, server return a response, and then stream is broken
2. Client sends `WaitExecution()`, server returns a few `Operation` messages, and then return a `NOT_FOUND` error
3. GOTO 1.

However, this is fixed by having two types of backoff:
1. `executeBackoff`: used for count the retry times for `Execute()` calls and can't be reset.
2. waitExecutionBackoff: used for count retry times for `WaitExecution()` calls and will be reset once we received a GOOD response from server.

Test case `executeRemotely_notFoundLoop_reportError` cover this case.